### ISSUE TITLE
add Update<R> to exercise and create method result types in Java codegen

### ISSUE THREAD

**Issue by S11001001** [digital-asset/daml#14312]
### What is the problem you want to solve?

Decoding is typeclass-driven in Scala codegen, so it is possible for users to put together type-safe utilities for working with choice results, like this one:

```scala
  import com.daml.ledger.client.{binding => sb}, sb.Primitive.Update
  import com.daml.ledger.api.v1.command_service.SubmitAndWaitRequest

  /** Assuming `ua` is an exercise from Scala codegen, submit, wait, and decode
    * the result according to the declared return type of the choice.
    */
  def performUpdate[A: sb.ValueDecoder](ledger: LedgerClient, ua: Update[A]): Future[Option[A]] =
    ledger.commandServiceClient
      .submitAndWaitForTransactionTree(
        // elided other details of the request
        SubmitAndWaitRequest(Some(Commands(commands = Seq(ua.command)))),
        None,
      )
      .map { swttr =>
        for {
          tree <- swttr.transaction
          exerciseEv <- tree.rootEventIds.collectFirst(Function unlift { id =>
            tree.eventsById.get(id).flatMap(_.kind.exercised)
          })
          resultValue <- exerciseEv.exerciseResult
          decoded <- sb.Value.decode(resultValue)
        } yield decoded
      }
```

Exercise methods in Scala codegen return `Update[R]` where `R` is the choice's return type, so the correct `ValueDecoder` instance can be implicitly selected when using such a utility. Moreover, we can inspect the command, and choose a value to decode according to whether it is a create, exercise, or create-and-exercise, giving us a full natural transformation from `Update[R]` to `Future[F[R]]` for some functor `F` (`Option` used above for demo purposes.).

On the other hand, exercise methods in Java codegen return `ExerciseCommand`, `ExerciseByKeyCommand`, or such, with no indication of what the choice return type will be.  Even if a phantom type parameter was added just to place the information somewhere, users cannot look up value-to-Java-codegen decoders implicitly in Java.

### What is the solution you would propose?

Introduce an `Update<R>` class that all generated `exercise*` methods will return.

```java
public Update<IouTransfer.ContractId> exerciseIou_Transfer(...) {...}
```

`Update` is a sum type of possible commands, as it effectively is in Scala codegen. This is a breaking change for existing callers that expect specific `*Command` types to be returned. Static evidence that your `exercise` call returned a particular subtype of command is significantly less valuable than static evidence of the Daml result of executing that command.

For `create`, we should be able to return `CreateUpdate<ContractId>`, where `CreateUpdate<Z> <: Update<Z> & CreateCommand`, to preserve compatibility.

`Update<R>` will carry a #14313 decoder for `R`, supplied by the codegenned exercise methods.

### Describe alternatives you've considered

1. Have users assemble and pass in a decoder for `R` themselves. This will simplify some aspects of the exercise method generation, but it is significantly harder for users to do that, especially since we already have the tools in the codegen implementation to assemble decoders for all possible `R`s that we can emit, and also given that we do not have a convenient combinator language for assembling decoders, i.e. #14313 (notwithstanding `ContractCompanion`, which is currently specific to templates).
2. Leave the type parameter off of `Update`, introducing a method that decodes to `java.lang.Object` which is "guaranteed" to be a value decoded to codegen. This allows us to avoid changing the return types at all. I elide description of the downside here.
3. Change the definition of Java to support Scala-like intersection types as method result types, so that we can return `Update<R> & Cmd`. This will allow existing code that statically expects `ExerciseCommand` _et al_ to continue to do so. Codegen users would have to upgrade to the earliest version of Java that supports this, though. Additionally, it might take some time to get this feature through the Java feature pipeline.

### Additional context

Inspired by [a forum discussion](https://discuss.daml.com/t/extracting-choice-return-type-defined-in-daml-templates-as-a-java-type/4666) with @juliuswilliam-da (though I do not think this would have genuinely applied to that use case). It has come up elsewhere on the forum, such as [here](https://discuss.daml.com/t/how-to-get-command-return-value-s-when-using-java-bindings/3879).

This would permit natural transformation utilities like `performUpdate` to be written for Java codegen users. In my opinion, the inability to conveniently, type-safely extract exercise results (and well-typed create-contract-IDs) as you can in the Daml surface language is a significant barrier to e.g. writing "triggers" in Java instead of Daml, mentioned by @garyverhaegen-da yesterday. We can proof-of-concept such utilities with this feature, but the real value is making them practical to write at all.

It would be possible to add `map2`, `pure`, and `map` for `Update` to support submitting multiple commands in a single submit. This would significantly complicate the aforementioned utilities, but require no further codegen changes.

We would prefer to return `Update<Z> & Cmd`, but Java doesn't have lower bounds on type parameters. Oh well!

**Comment by S11001001** [comment]
For comparison, consider exercise in TypeScript bindings:

https://github.com/digital-asset/daml/blob/1ab1605948f471d87eb17fcb6bc4f86d056485b3/language-support/ts/daml-ledger/index.ts#L1104-L1108

Here `R` represents the decoded return value from exercise; rather than simply producing an `ExerciseCommand`, this utility produces the command, submits it, and applies the appropriate typed decoder to the result, so the user immediately has it at hand. Part of accomplishing this is having a TypeScript equivalent to #14313.

**Comment by chunlokling-da** [comment]
Based on the details in the ticket, Here is the summeary the options:
(1) To use intersection types so that we can return Update<R> & Cmd
(2) maybe users simply in practice treat them as Command just treat all these as Exercise so just change the return type to Update<R>
(3)newly mangled method names
IMO I am leaning towards 2 but it really depends on how users use the generated code.

Why is it so undesirable to have (3) ? (create a parallel set of methods with mangled names that return Update<R> and deprecate those returning Exercise<> Cmd )?  I think It will still be a good intermediate step even though ultimately we don't want to have it.

**Comment by chunlokling-da** [comment]
@S11001001 mentioned: if users don’t care, we do (2) they have to do nothing, we do (3) they have to change literally every exercise

**Comment by chunlokling-da** [comment]
So the main question is how users use these exercise* methods that return ExerciseCommand? Do they actually use the specific Command?

If we can get an answer to these questions we can make a decision on whether (2) just change the return type on existing methods or (3) have a new set of exercise methods with managed names.

@stefanobaghino-da @ray-roestenburg-da  I wonder if it is possible to get a clear picture of how our users of codegen use these methods? There might be public developers who are using codegen which we are not able to investigate how they use.

In the case that it is tricky to get a clear picture of how codegen users (both our clients and public users) use those methods. I would think (3) having a new set of exercise methods with managed names will be safest.

**Comment by ray-roestenburg-da** [comment]
@chunlokling-da option 3 is only an option if we can come up with better names than the ones we have already, and that looks to be very hard. Coming up with some 'synthetic', worse names, and deprecate the better fitting names is the opposite we would want in the long run. We need to be able to make changes to generated code that improve the user experience in general. Please go with option 2. As @S11001001 mentioned, it is possible that many users don't notice, and fixes should be straightforward.

**Comment by S11001001** [comment]
Implementation notes:

* [ ] depend on #15116 (and its interface counterpart if a separate PR)
* [ ] add a result value decoder to `ChoiceMetadata`
* [ ] change `makeExerciseCmd`'s signature to `<T, R> Update<R> (ChoiceMetadata<?, ? super T, ? extends R> choice, T choiceArgument)`
* [ ] retrieve that decoder from the `CHOICE_x` available locally, for inclusion in the `makeExerciseCmd` call

**Comment by cocreature** [comment]
Not quite sure if that’s what you’re looking for but we heavily use the `Update<R>` types with the Scala codegen to define a `submitWithResult` that exercises a single command and gives us back the contract id for creates and the exercise result for exercises:
```
  def submitWithResult[T](
      actAs: Seq[PartyId],
      readAs: Seq[PartyId],
      update: P.Update[T],
      commandId: Option[String] = None,
      deduplicationTime: Option[NonNegativeFiniteDuration] = None,
  )(implicit traceContext: TraceContext, decoder: ValueDecoder[T]): Future[T]
```
Being able to define some equivalent of that using the Java codegen would be great.

**Comment by S11001001** [comment]
@cocreature Yes, we are interested exactly in such use-cases. Specifically, we want to let you do that (and effectively `ValueDecoder` will be packaged up with the `Update` in Java) by sacrificing the static `ExerciseCommand` return type from exercise methods, and likely similarly the static `CreateCommand`, as `Update` will become a sum command type similar to the `Command` member in `P.Update`.

But as a result we are wondering whether anyone cares about those static result command types. They obviously do not in Scala codegen, because it has never supported them, but 🤷

**Comment by cocreature** [comment]
Yeah I'm not aware of any case where they would help us.

**Comment by juliuswilliam-da** [comment]
Thanks @stephencompall-DA for the prod, apologies I haven't been following up on this discussion.
The request to return the appropriate daml type originated due to a client requirement to create a JSON API endpoint documentation. 
By having the return type provided in the Java codegen, we are able to utilize the return type to generate a swagger documentation which indicates the return object of each exercise endpoint.

@cocreature That sounds exactly like what we're looking for with the Java codegen. In the scenario of submitWithResult, does an actual exercise have to be submitted to the ledger to retrieve the exercise result for that choice?

**Comment by cocreature** [comment]
> @cocreature That sounds exactly like what we're looking for with the Java codegen. In the scenario of submitWithResult, does an actual exercise have to be submitted to the ledger to retrieve the exercise result for that choice?

Yes

**Comment by juliuswilliam-da** [comment]
> > I wonder if it is possible to get a clear picture of how our users of codegen use these methods?
> 
> @juliuswilliam-da How do you use the result values returned from `exercise*` methods generated by Java codegen? Do you pass them directly to a specific Java bindings method? Convert immediately to protobuf? Or perhaps inspect their contents directly for some other use?

We will use a script to parse the return type string in the Java file and create an annotated Java class file. This annotated file is then used by the Swagger generator library to create a yaml file with the appropriate endpoints and return type. 

This return type doesn't have to strictly be done within the Java codegen. If there is some sort of daml doc parser for any of the supported output doc templates (rst, md, markdown,html, hoogle, json) which contains a mapping of ChoiceName : ChoiceReturnType, we can manually parse that instead of referring to the generated java codegen files for extracting the return type as well since all this parsing is done line-by-line via an external script before being used to construct the java class files for swagger doc generation.

**Comment by cocreature** [comment]
I think I'd just parse daml-lf for that instead of trying to generated java code.
